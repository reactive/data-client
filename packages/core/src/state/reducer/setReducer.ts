import { normalize } from '@data-client/normalizr';

import Controller from '../../controller/Controller.js';
import type { State, SetAction } from '../../types.js';

export function setReducer(
  state: State<unknown>,
  action: SetAction,
  controller: Controller,
) {
  let value: any;
  if (typeof action.value === 'function') {
    const previousValue = controller.get(action.schema, ...action.args, state);
    if (previousValue === undefined) return state;
    value = action.value(previousValue);
  } else {
    value = action.value;
  }
  try {
    const { entities, indexes, entitiesMeta } = normalize(
      action.schema,
      value,
      action.args,
      state,
      action.meta,
    );
    return {
      entities,
      endpoints: state.endpoints,
      indexes,
      meta: state.meta,
      entitiesMeta,
      optimistic: state.optimistic,
      lastReset: state.lastReset,
    };
    // reducer must update the state, so in case of processing errors we simply compute the endpoints inline
  } catch (error: any) {
    // this is not always bubbled up, so let's double sure this doesn't fail silently
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      console.error(error);
    }
    return state;
  }
}
