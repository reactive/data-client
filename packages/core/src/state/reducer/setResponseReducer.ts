import { normalize } from '@data-client/normalizr';

import { OPTIMISTIC } from '../../actionTypes.js';
import AbortOptimistic from '../../controller/AbortOptimistic.js';
import type Controller from '../../controller/Controller.js';
import type {
  State,
  SetResponseAction,
  OptimisticAction,
} from '../../types.js';

export function setResponseReducer(
  state: State<unknown>,
  action: OptimisticAction | SetResponseAction,
  controller: Controller,
) {
  if (action.error) {
    return reduceError(state, action, action.response);
  }
  try {
    let response: any;
    // for true set's response is contained in action
    if (action.type === OPTIMISTIC) {
      // this should never happen
      /* istanbul ignore if */
      if (!action.endpoint.getOptimisticResponse) return state;
      try {
        // compute optimistic response based on current state
        response = action.endpoint.getOptimisticResponse.call(
          action.endpoint,
          controller.snapshot(state, action.meta.fetchedAt),
          ...action.args,
        );
      } catch (e: any) {
        // AbortOptimistic means 'do nothing', otherwise we count the exception as endpoint failure
        if (e.constructor === AbortOptimistic) {
          return state;
        }
        throw e;
      }
    } else {
      response = action.response;
    }
    const { result, entities, indexes, entitiesMeta } = normalize(
      action.endpoint.schema,
      response,
      action.args,
      state,
      action.meta,
    );
    const endpoints: Record<string, unknown> = {
      ...state.endpoints,
      [action.key]: result,
    };
    try {
      if (action.endpoint.update) {
        const updaters = action.endpoint.update(result, ...action.args);
        Object.keys(updaters).forEach(key => {
          endpoints[key] = updaters[key](endpoints[key]);
        });
      }
      // no reason to completely fail because of user-code error
      // integrity of this state update is still guaranteed
    } catch (error) {
      console.error(
        `The following error occured during Endpoint.update() for ${action.key}`,
      );
      console.error(error);
    }
    return {
      entities,
      endpoints,
      indexes,
      meta: {
        ...state.meta,
        [action.key]: {
          date: action.meta.date,
          fetchedAt: action.meta.fetchedAt,
          expiresAt: action.meta.expiresAt,
          prevExpiresAt: state.meta[action.key]?.expiresAt,
        },
      },
      entitiesMeta,
      optimistic: filterOptimistic(state, action),
      lastReset: state.lastReset,
    };
    // reducer must update the state, so in case of processing errors we simply compute the endpoints inline
  } catch (error: any) {
    if (typeof error === 'object') {
      error.message = `Error processing ${
        action.key
      }\n\nFull Schema: ${JSON.stringify(
        action.endpoint.schema,
        undefined,
        2,
      )}\n\nError:\n${error.message}`;
      if ('response' in action) error.response = action.response;
      error.status = 400;
    }

    // this is not always bubbled up, so let's double sure this doesn't fail silently
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      console.error(error);
    }
    return reduceError(state, action, error);
  }
}

function reduceError(
  state: State<unknown>,
  action: SetResponseAction | OptimisticAction,
  error: any,
): State<unknown> {
  if (error.name === 'AbortError') {
    // In case we abort simply undo the optimistic update and act like no fetch even occured
    // We still want those watching promises from fetch directly to observed the abort, but we don't want to
    // Trigger errors in this case. This means theoretically improperly built abortes useResource() could suspend forever.
    return {
      ...state,
      optimistic: filterOptimistic(state, action),
    };
  }
  return {
    ...state,
    meta: {
      ...state.meta,
      [action.key]: {
        date: action.meta.date,
        fetchedAt: action.meta.fetchedAt,
        expiresAt: action.meta.expiresAt,
        error,
        errorPolicy: action.endpoint.errorPolicy?.(error),
      },
    },
    optimistic: filterOptimistic(state, action),
  };
}
/** Filter all requests with same serialization that did not start after the resolving request */
function filterOptimistic(
  state: State<unknown>,
  resolvingAction: SetResponseAction | OptimisticAction,
) {
  return state.optimistic.filter(
    optimisticAction =>
      optimisticAction.key !== resolvingAction.key ||
      (optimisticAction.type === OPTIMISTIC ?
        optimisticAction.meta.fetchedAt !== resolvingAction.meta.fetchedAt
      : optimisticAction.meta.date > resolvingAction.meta.date),
  );
}
