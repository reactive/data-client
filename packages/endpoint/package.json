{
  "name": "@data-client/endpoint",
  "version": "0.14.17",
  "description": "Declarative Network Interface Definitions",
  "homepage": "https://dataclient.io/docs/guides/custom-protocol",
  "keywords": [
    "data",
    "cache",
    "networking",
    "fetch",
    "typescript",
    "data fetching",
    "data cache",
    "api",
    "normalized cache",
    "subject pattern",
    "declarative",
    "definition",
    "schema",
    "async",
    "ajax",
    "react",
    "front-end",
    "mobile",
    "react-native",
    "ios",
    "android",
    "web",
    "expo",
    "async",
    "promise"
  ],
  "sideEffects": false,
  "main": "dist/index.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=5.4": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=4.8": {
      "": [
        "ts4.8/index.d.ts"
      ],
      "*": [
        "ts4.8/index.d.ts"
      ]
    },
    ">=4.2": {
      "": [
        "ts4.2/index.d.ts"
      ],
      "*": [
        "ts4.2/index.d.ts"
      ]
    },
    ">=4.0": {
      "": [
        "ts4.0/index.d.ts"
      ],
      "*": [
        "ts4.0/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "ts4.8",
    "ts4.2",
    "ts4.0",
    "ts3.4",
    "node.mjs",
    "legacy",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib --ignore 'src/schemas-3.7'",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy --ignore 'src/schemas-3.7'",
    "build:js:node": "BROWSERSLIST_ENV=node12 yarn g:rollup && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:js:browser": "BROWSERSLIST_ENV=legacy yarn g:rollup",
    "build:bundle": "yarn g:runs build:js:\\*",
    "build:clean": "yarn g:clean ts3.4 ts4.0 ts4.2 ts4.8",
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "build:legacy-types": "yarn g:legacy-types 4.8 4.2 4.0 3.4",
    "dev": "run build:lib -w",
    "prepare": "run build:lib",
    "prepack": "run prepare && run build:bundle && run build:legacy:lib",
    "tsc:ci": "yarn g:tsc --project tsconfig.test.json",
    "typecheck": "run tsc:ci"
  },
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/endpoint"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "dependencies": {
    "@babel/runtime-corejs3": "^7.26.0",
    "core-js-pure": "^3.40.0"
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/core": "workspace:*",
    "@data-client/normalizr": "workspace:*",
    "@js-temporal/polyfill": "^0.4.4",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "immutable": "4.2.2",
    "nock": "13.3.1",
    "rollup-plugins": "workspace:*"
  }
}
