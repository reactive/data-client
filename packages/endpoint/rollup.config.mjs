import {
  babel,
  dts,
  commonjs,
  filesize,
  json,
  resolve,
  replace,
  terser,
  onwarn,
  resolveTsAsJs,
} from 'rollup-plugins';

import pkg from './package.json' with { type: 'json' };

const dependencies = Object.keys(pkg.dependencies).filter(
  dep => !['@babel/runtime'].includes(dep),
);

const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node'];
const nativeExtensions = ['.native.ts', ...extensions];
process.env.NODE_ENV = 'production';

function isExternal(id) {
  return dependencies.some(dep => dep === id || id.startsWith(dep));
}

const configs = [];
if (process.env.BROWSERSLIST_ENV !== 'node12') {
  // browser-friendly UMD build
  configs.push({
    input: 'lib/index.js',
    output: [{ file: pkg.unpkg, format: 'umd', name: 'Endpoint' }],
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        rootMode: 'upward',
        babelHelpers: 'runtime',
        caller: { polyfillMethod: false },
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('production'),
        preventAssignment: true,
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      json(),
      terser({}),
      filesize({ showBrotliSize: true }),
    ],
  });
  configs.push({
    input: './lib/index.d.ts',
    output: [{ file: 'index.d.ts', format: 'es' }],
    plugins: [dts()],
  });
} else {
  // node-friendly commonjs build
  configs.push({
    input: 'src/index.ts',
    external: isExternal,
    output: [{ file: pkg.main, format: 'cjs' }],
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '**/__tests__/**', '**/*.d.ts'],
        extensions: nativeExtensions,
        rootMode: 'upward',
        babelHelpers: 'runtime',
      }),
      replace({ 'process.env.CJS': 'true', preventAssignment: true }),
      resolve({ extensions: nativeExtensions }),
      resolveTsAsJs,
      commonjs({ extensions: nativeExtensions }),
    ],
  });
}
export default configs;
