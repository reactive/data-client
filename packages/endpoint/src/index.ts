export type {
  EndpointOptions,
  EndpointInstance,
  EndpointInstanceInterface,
  EndpointExtendOptions,
} from './endpoint.js';
export * as schema from './schema.js';
// Without this we get 'cannot be named without a reference to' for resource()....why is this?
// Clue 1) It only happens with types mentioned in return types of other types
export type { Array, Invalidate, Collection, DefaultArgs } from './schema.js';
export { default as Entity } from './schemas/Entity.js';
export { default as EntityMixin } from './schemas/EntityMixin.js';
export { default as validateRequired } from './validateRequired.js';
export { INVALID } from './special.js';
export type {
  EndpointInterface,
  ReadEndpoint,
  MutateEndpoint,
  Schema,
  SnapshotInterface,
  ExpiryStatusInterface,
  SchemaSimple,
  SchemaClass,
  PolymorphicInterface,
  Queryable,
} from './interface.js';
export type { EntityFields } from './schemas/EntityFields.js';
export type {
  AbstractInstanceType,
  Normalize,
  NormalizeNullable,
  Denormalize,
  DenormalizeNullable,
  SchemaArgs,
  ObjectArgs,
  EntityMap,
  RecordClass,
  NormalizedNullableObject,
  NormalizeObject,
  NormalizedEntity,
  DenormalizeObject,
  DenormalizeNullableObject,
} from './normal.js';
export type {
  EndpointExtraOptions,
  FetchFunction,
  ResolveType,
  EndpointParam,
  NetworkError,
  UnknownError,
  ErrorTypes,
  EndpointToFunction,
} from './types.js';
export type { NI } from './NoInfer.js';

export { default as Endpoint, ExtendableEndpoint } from './endpoint.js';
export type { KeyofEndpointInstance } from './endpoint.js';
