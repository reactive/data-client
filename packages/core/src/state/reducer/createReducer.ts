import { expireReducer } from './expireReducer.js';
import { fetchReducer } from './fetchReducer.js';
import { invalidateReducer } from './invalidateReducer.js';
import { setReducer } from './setReducer.js';
import { setResponseReducer } from './setResponseReducer.js';
import {
  SET,
  INVALIDATE,
  RESET,
  FETCH,
  GC,
  OPTIMISTIC,
  INVALIDATEALL,
  EXPIREALL,
  SET_RESPONSE,
} from '../../actionTypes.js';
import type Controller from '../../controller/Controller.js';
import type { ActionTypes, State } from '../../types.js';

export default function createReducer(controller: Controller): ReducerType {
  return function reducer(
    state: State<unknown> | undefined,
    action: ActionTypes,
  ): State<unknown> {
    if (!state) state = initialState;
    switch (action.type) {
      case GC:
        // inline deletes are fine as these should have 0 refcounts
        action.entities.forEach(({ key, pk }) => {
          delete (state as any).entities[key]?.[pk];
          delete (state as any).entitiesMeta[key]?.[pk];
        });
        action.endpoints.forEach(fetchKey => {
          delete (state as any).endpoints[fetchKey];
          delete (state as any).meta[fetchKey];
        });
        return state;
      case FETCH:
        return fetchReducer(state, action);

      case OPTIMISTIC:
      // eslint-disable-next-line no-fallthrough
      case SET_RESPONSE:
        return setResponseReducer(state, action, controller);

      case SET:
        return setReducer(state, action, controller);

      case INVALIDATEALL:
      case INVALIDATE:
        return invalidateReducer(state, action);

      case EXPIREALL:
        return expireReducer(state, action);

      case RESET:
        return { ...initialState, lastReset: action.date };

      default:
        // A reducer must always return a valid state.
        // Alternatively you can throw an error if an invalid action is dispatched.
        return state;
    }
  } as any;
}

export const initialState: State<unknown> = {
  entities: {},
  endpoints: {},
  indexes: {},
  meta: {},
  entitiesMeta: {},
  optimistic: [],
  lastReset: 0,
};

type ReducerType = (
  state: State<unknown> | undefined,
  action: ActionTypes,
) => State<unknown>;

type Writable<T> = { [P in keyof T]: NonNullable<Writable<T[P]>> };
