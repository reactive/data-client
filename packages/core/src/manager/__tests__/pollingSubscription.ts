import { Endpoint } from '@data-client/endpoint';
import { Article } from '__tests__/new';

import { createSubscription } from '../../controller/actions/createSubscription';
import Controller from '../../controller/Controller';
import { initialState } from '../../state/reducer/createReducer';
import ConnectionListener from '../ConnectionListener';
import DefaultConnectionListener from '../DefaultConnectionListener';
import PollingSubscription from '../PollingSubscription';

class MockConnectionListener implements ConnectionListener {
  declare online: boolean;
  declare onlineHandlers: (() => void)[];
  declare offlineHandlers: (() => void)[];

  constructor(online: boolean) {
    this.online = online;
    this.onlineHandlers = [];
    this.offlineHandlers = [];
  }

  isOnline() {
    return this.online;
  }

  addOnlineListener(handler: () => void) {
    this.onlineHandlers.push(handler);
  }

  removeOnlineListener(handler: () => void) {
    this.onlineHandlers = this.onlineHandlers.filter(h => h !== handler);
  }

  addOfflineListener(handler: () => void) {
    this.offlineHandlers.push(handler);
  }

  removeOfflineListener(handler: () => void) {
    this.offlineHandlers = this.offlineHandlers.filter(h => h !== handler);
  }

  trigger(event: 'offline' | 'online') {
    switch (event) {
      case 'offline':
        this.online = false;
        this.offlineHandlers.forEach(t => t());
        break;
      case 'online':
        this.online = true;
        this.onlineHandlers.forEach(t => t());
        break;
    }
  }

  reset() {
    this.offlineHandlers = [];
    this.onlineHandlers = [];
  }
}

const makeState = (key: string, time: number) => () => ({
  ...initialState,
  meta: { [key]: { date: time, expiresAt: Infinity } },
});

function onError(e: any) {
  e.preventDefault();
}
beforeEach(() => {
  if (typeof addEventListener === 'function')
    addEventListener('error', onError);
});
afterEach(() => {
  if (typeof removeEventListener === 'function')
    removeEventListener('error', onError);
});

describe('PollingSubscription', () => {
  const a = () => Promise.resolve({ id: 5, title: 'hi' });
  const fetch = jest.fn(a);
  const endpoint = new Endpoint(fetch, {
    schema: Article,
    key: () => 'test.com',
    pollFrequency: 5000,
  });

  describe('existing data', () => {
    it('should wait to call for fresh data', () => {
      const dispatch = jest.fn();
      const controller = new Controller({ dispatch });
      (controller as any).getState = makeState('test.com', Date.now());

      jest.useFakeTimers();
      const sub2 = new PollingSubscription(
        createSubscription(endpoint, { args: [] }),
        controller,
      );
      expect(dispatch.mock.calls.length).toBe(0);
      jest.advanceTimersByTime(4990);
      expect(dispatch.mock.calls.length).toBe(0);
      jest.advanceTimersByTime(20);
      expect(dispatch.mock.calls.length).toBe(1);
      jest.advanceTimersByTime(5000);
      expect(dispatch.mock.calls.length).toBe(2);
      jest.advanceTimersByTime(2000);
      expect(dispatch.mock.calls.length).toBe(2);
      sub2.cleanup();
      jest.useRealTimers();
    });

    it('should only run once with multiple simultaneous starts', () => {
      const dispatch = jest.fn();
      const controller = new Controller({ dispatch });
      (controller as any).getState = () => initialState;

      jest.useFakeTimers();
      const sub2 = new PollingSubscription(
        createSubscription(endpoint, { args: [] }),
        controller,
      );
      sub2.add(1000);
      sub2.add(1000);
      sub2.add(1000);
      jest.advanceTimersByTime(1);
      expect(dispatch.mock.calls.length).toBe(1);
      jest.advanceTimersByTime(999);
      expect(dispatch.mock.calls.length).toBe(2);
      jest.advanceTimersByTime(10);
      sub2.remove(1000);
      sub2.remove(1000);
      sub2.remove(1000);
      sub2.cleanup();
      jest.useRealTimers();
    });
  });

  describe('fresh data', () => {
    const dispatch = jest.fn();
    const controller = new Controller({ dispatch });
    (controller as any).getState = makeState('test.com', 0);
    let sub: PollingSubscription;

    beforeAll(() => {
      jest.useFakeTimers();
      sub = new PollingSubscription(
        createSubscription(endpoint, { args: [] }),
        controller,
      );
    });
    afterAll(() => {
      sub.cleanup();
      jest.useRealTimers();
    });

    it('should throw on undefined frequency in construction', () => {
      expect(
        () =>
          new PollingSubscription(
            createSubscription(
              new Endpoint(fetch, {
                schema: Article,
                key: () => 'test.com',
              }),
              { args: [] },
            ),
            controller,
          ),
      ).toThrow();
    });

    it('should call immediately', () => {
      jest.advanceTimersByTime(1);
      expect(dispatch.mock.calls.length).toBe(1);
    });

    it('should call after period', () => {
      dispatch.mockReset();
      jest.advanceTimersByTime(5000);
      expect(dispatch.mock.calls.length).toBe(1);
      dispatch.mock.calls[0].forEach((element: any) => {
        delete element?.meta?.fetchedAt;
      });
      expect(dispatch.mock.calls[0]).toMatchSnapshot();
      jest.advanceTimersByTime(5000);
      expect(dispatch.mock.calls.length).toBe(2);
      dispatch.mock.calls[1].forEach((element: any) => {
        delete element?.meta?.fetchedAt;
      });

      expect(dispatch.mock.calls[1]).toMatchSnapshot();
    });

    describe('add()', () => {
      it('should take smaller frequency when another is added', () => {
        sub.add(1000);
        jest.advanceTimersByTime(1000 * 4);
        expect(dispatch.mock.calls.length).toBe(2 + 4);
      });

      it('should not change frequency if same is added', () => {
        dispatch.mockClear();
        sub.add(1000);
        jest.advanceTimersByTime(1000 * 13);
        expect(dispatch.mock.calls.length).toBe(13);
      });

      it('should not change frequency if larger is added', () => {
        dispatch.mockClear();
        sub.add(7000);
        sub.add(8000);
        jest.advanceTimersByTime(1000 * 13);
        expect(dispatch.mock.calls.length).toBe(13);
      });

      it('should do nothing if frequency is undefined', () => {
        dispatch.mockClear();
        sub.add(undefined);
        jest.advanceTimersByTime(1000 * 13);
        expect(dispatch.mock.calls.length).toBe(13);
      });
    });

    describe('remove()', () => {
      it('should not change frequency if only partially removed', () => {
        dispatch.mockClear();
        sub.remove(1000);
        jest.advanceTimersByTime(1000 * 13);
        expect(dispatch.mock.calls.length).toBe(13);
      });

      it('should not change frequency if only larger removed', () => {
        dispatch.mockClear();
        sub.remove(7000);
        jest.advanceTimersByTime(1000 * 13);
        expect(dispatch.mock.calls.length).toBe(13);
      });

      it('should go back to fastest if smallest frequency is removed completely', () => {
        sub.remove(1000);
        jest.advanceTimersByTime(1000);
        dispatch.mockClear();
        jest.advanceTimersByTime(5000 * 13);
        expect(dispatch.mock.calls.length).toBe(13);
      });

      it('should do nothing if frequency is not registered', () => {
        const oldError = console.error;
        const spy = (console.error = jest.fn());

        sub.remove(1000);
        dispatch.mockClear();
        jest.advanceTimersByTime(5000 * 13);
        expect(dispatch.mock.calls.length).toBe(13);

        expect(spy.mock.calls[0]).toMatchInlineSnapshot(`
          [
            "Mismatched remove: 1000 is not subscribed for test.com",
          ]
        `);
        console.error = oldError;
      });

      it('should do nothing if frequency is undefined', () => {
        sub.remove(undefined);
        dispatch.mockClear();
        jest.advanceTimersByTime(5000 * 13);
        expect(dispatch.mock.calls.length).toBe(13);
      });

      it('should return false until completely empty, then return true', () => {
        const controller = new Controller({ dispatch });
        (controller as any).getState = makeState('test.com', 0);

        const sub = new PollingSubscription(
          createSubscription(endpoint, { args: [] }),
          controller,
        );
        sub.add(5000);
        sub.add(7000);
        expect(sub.remove(5000)).toBe(false);
        expect(sub.remove(5000)).toBe(false);
        expect(sub.remove(7000)).toBe(true);
      });
    });

    describe('cleanup()', () => {
      let warnSpy: jest.Spied<typeof console.warn>;
      afterEach(() => {
        warnSpy.mockRestore();
      });
      beforeEach(() => {
        (warnSpy = jest.spyOn(console, 'warn')).mockImplementation(() => {});
      });

      it('should stop all timers', () => {
        dispatch.mockClear();
        sub.cleanup();
        jest.advanceTimersByTime(5000 * 13);
        expect(dispatch.mock.calls.length).toBe(0);
      });
      it('should be idempotent', () => {
        sub.cleanup();
      });
      it('should not run even if interval not cancelled', () => {
        const controller = new Controller({ dispatch });
        (controller as any).getState = makeState('test.com', 0);

        sub.cleanup();
        sub = new PollingSubscription(
          createSubscription(endpoint.extend({ key: () => 'test.com2' }), {
            args: [],
          }),
          controller,
        );
        sub.add(5000);
        jest.runOnlyPendingTimers();
        delete (sub as any).intervalId;
        jest.advanceTimersByTime(5000 * 13);

        expect(dispatch.mock.calls.length).toBe(1);
        expect(warnSpy.mock.calls.length).toBe(13);
        expect(warnSpy.mock.calls[0]).toMatchSnapshot();
      });
    });
  });

  describe('offline support', () => {
    beforeAll(() => {
      jest.useFakeTimers();
    });
    afterAll(() => {
      jest.useRealTimers();
    });

    function createMocks(listener: ConnectionListener) {
      const dispatch = jest.fn();
      const a = () => Promise.resolve({ id: 5, title: 'hi' });
      const fetch = jest.fn(a);
      const endpoint = new Endpoint(fetch, {
        schema: Article,
        key: () => 'test.com',
        pollFrequency: 5000,
      });
      const controller = new Controller({ dispatch });
      (controller as any).getState = makeState('test.com', 0);

      const pollingSubscription = new PollingSubscription(
        createSubscription(endpoint, {
          args: [],
        }),
        controller,
        listener,
      );
      jest.advanceTimersByTime(1);
      return { dispatch, fetch, pollingSubscription };
    }

    it('should not dispatch when offline', () => {
      jest.spyOn(navigator, 'onLine', 'get').mockReturnValue(false);
      const listener = new DefaultConnectionListener();
      const offlineSpy = jest.spyOn(listener, 'addOfflineListener');
      const onlineSpy = jest.spyOn(listener, 'addOnlineListener');
      const { dispatch } = createMocks(listener);
      jest.advanceTimersByTime(50000);
      jest.spyOn(navigator, 'onLine', 'get').mockReturnValue(true);
      expect(dispatch.mock.calls.length).toBe(0);
      expect(offlineSpy.mock.calls.length).toBe(0);
      expect(onlineSpy.mock.calls.length).toBe(1);
    });

    it('should not dispatch when onLine is undefined (default to true)', () => {
      jest.spyOn(navigator, 'onLine', 'get').mockReturnValue(undefined as any);
      const listener = new DefaultConnectionListener();
      const offlineSpy = jest.spyOn(listener, 'addOfflineListener');
      const onlineSpy = jest.spyOn(listener, 'addOnlineListener');
      const { dispatch } = createMocks(listener);
      expect(dispatch.mock.calls.length).toBe(1);
      expect(offlineSpy.mock.calls.length).toBe(1);
      expect(onlineSpy.mock.calls.length).toBe(0);
    });

    it('should immediately start fetching when online', () => {
      const listener = new MockConnectionListener(false);
      const { dispatch } = createMocks(listener);
      expect(dispatch.mock.calls.length).toBe(0);

      listener.trigger('online');
      jest.advanceTimersByTime(1);
      expect(dispatch.mock.calls.length).toBe(1);
      jest.advanceTimersByTime(5000);
      expect(dispatch.mock.calls.length).toBe(2);
      expect(listener.offlineHandlers.length).toBe(1);
      expect(listener.onlineHandlers.length).toBe(0);
    });

    it('should not run when timeoutId is deleted after coming online', () => {
      const listener = new MockConnectionListener(false);
      const { dispatch, pollingSubscription } = createMocks(listener);
      expect(dispatch.mock.calls.length).toBe(0);

      listener.trigger('online');
      delete (pollingSubscription as any).startId;
      jest.advanceTimersByTime(1);
      expect(dispatch.mock.calls.length).toBe(0);
      jest.advanceTimersByTime(5000);
      expect(dispatch.mock.calls.length).toBe(0);
      expect(listener.offlineHandlers.length).toBe(1);
      expect(listener.onlineHandlers.length).toBe(0);
    });

    it('should stop dispatching when offline again', () => {
      const listener = new MockConnectionListener(true);
      const { dispatch } = createMocks(listener);
      expect(dispatch.mock.calls.length).toBe(1);

      listener.trigger('offline');
      jest.advanceTimersByTime(50000);
      expect(dispatch.mock.calls.length).toBe(1);
      expect(listener.offlineHandlers.length).toBe(0);
      expect(listener.onlineHandlers.length).toBe(1);
    });
  });
});
