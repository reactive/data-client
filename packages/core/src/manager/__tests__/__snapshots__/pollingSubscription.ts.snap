// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PollingSubscription fresh data cleanup() should not run even if interval not cancelled 1`] = `
[
  "Poll intervalId for test.com2 still running, but intervalId deleted",
]
`;

exports[`PollingSubscription fresh data should call after period 1`] = `
[
  {
    "args": [],
    "endpoint": [Function],
    "key": "test.com",
    "meta": {
      "promise": Promise {},
      "reject": [Function],
      "resolve": [Function],
    },
    "type": "rdc/fetch",
  },
]
`;

exports[`PollingSubscription fresh data should call after period 2`] = `
[
  {
    "args": [],
    "endpoint": [Function],
    "key": "test.com",
    "meta": {
      "promise": Promise {},
      "reject": [Function],
      "resolve": [Function],
    },
    "type": "rdc/fetch",
  },
]
`;
